services/src/main/java/org/keycloak/authorization/authorization/AuthorizationTokenService.java

```diff
@@ -802,5 +806,100 @@ RealmModel getRealm() {
         ClientConnection getClientConnection() {
             return clientConnection;
         }
+
+        public void addPermissions(List<String> permissionList, String permissionResourceFormat) {
+            if (permissionResourceFormat == null) {
+                permissionResourceFormat = "id";
+            }
+
+            switch (permissionResourceFormat) {
+                case "id":
+                    addPermissionsById(permissionList);
+                    break;
+                case "uri":
+                    addPermissionsByUri(permissionList);
+                    break;
+                default:
+                    // return nothing
+                    return;
+            }
+
+        }
+
+        private void addPermissionsById(List<String> permissionList) {
+            for (String permission : permissionList) {
+                String[] parts = permission.split("#");
+                String rsid = parts[0];
+
+                if (parts.length == 1) {
+                    addPermission(rsid);
+                } else {
+                    String[] scopes = parts[1].split(",");
+                    addPermission(rsid, scopes);
+                }
+            }
+        }
+
+        private void addPermissionsByUri(List<String> permissionList) {
+            StoreFactory storeFactory = authorization.getStoreFactory();
+
+            for (String permission : permissionList) {
+                String[] parts = permission.split("#");
+                String uri = parts[0];
+
+                List<Resource> resources = getResourceListByUri(uri, storeFactory);
+
+                if (parts.length == 1) {
+                    resources.stream().forEach(resource -> addPermission(resource.getId()));
+                } else {
+                    String[] scopes = parts[1].split(",");
+                    resources.stream().forEach(resource -> addPermission(resource.getId(), scopes));
+                }
+            }
+        }
+
+        private List<Resource> getResourceListByUri(String uri, StoreFactory storeFactory) {
+            Map<Resource.FilterOption, String[]> search = new EnumMap<>(Resource.FilterOption.class);
+            search.put(Resource.FilterOption.URI, new String[] { uri });
+            ResourceServer resourceServer = storeFactory.getResourceServerStore()
+                .findByClient(getRealm().getClientByClientId(getAudience()));
+            List<Resource> resources = storeFactory.getResourceStore().find(getRealm(), resourceServer, search, -1,
+                Constants.DEFAULT_MAX_RESULTS);
+
+            if (!resources.isEmpty()) {
+                return resources;
+            }
+
```

Comment:
Why searching by `Resource.FilterOption.URI` is not enough?

Reply:
If a request using a wildcard makes `resources.isEmpry() == false`, how about judging whether a request includes a wild card at first and execute each matching process?

Reply:
@tnorimat 
Thanks. Path matching is not for requests using a wildcard but for resources using a wildcard defined by the authorization service.
BTW, we think it's better to be able to specify whether we'd like to use path matching or not the same as ResourceSetService, especially considering performance, I introduced the `permission_resource_matching_uri` parameter.